common/persistence/visibility/visibility_manager_dual.go

```diff
@@ -28,38 +28,83 @@ import (
 	"fmt"
 
 	"go.temporal.io/api/serviceerror"
-
 	"go.temporal.io/server/common/dynamicconfig"
 	"go.temporal.io/server/common/namespace"
 	"go.temporal.io/server/common/persistence/visibility/manager"
 )
 
 type (
 	visibilityManagerDual struct {
-		stdVisibilityManager          manager.VisibilityManager
-		advVisibilityManager          manager.VisibilityManager
+		stdVisibilityManager manager.VisibilityManager
+		advVisibilityManager manager.VisibilityManager
+		managerSelector      managerSelector
+	}
+
+	managerSelector interface {
+		readManager(namespace namespace.Name) manager.VisibilityManager
+		writeManagers() ([]manager.VisibilityManager, error)
+	}
+
+	sqlToESManagerSelector struct {
 		enableAdvancedVisibilityRead  dynamicconfig.BoolPropertyFnWithNamespaceFilter
 		advancedVisibilityWritingMode dynamicconfig.StringPropertyFn
+		stdVisibilityManager          manager.VisibilityManager
+		advVisibilityManager          manager.VisibilityManager
+	}
+
+	esManagerSelector struct {
```

Comment:
Please place all these structs and their methods in separate files. I hate to mix structs with methods in one file (I know we have many of those but it is just a legacy and I am constantly fixing it).